blockchain_publisher/src/main/java/org/cardanofoundation/lob/app/blockchain_publisher/repository/TransactionEntityRepositoryGateway.java

```diff
@@ -24,20 +28,41 @@
 @RequiredArgsConstructor
 @Slf4j
 @Transactional(readOnly = true)
+@Getter
 public class TransactionEntityRepositoryGateway {
 
     private final TransactionEntityRepository transactionEntityRepository;
     private final TransactionItemEntityRepository transactionItemEntityRepository;
 
+    @Value("${lob.blockchain_publisher.dispatcher.lock_timeout:PT3H}") // Default grace period to 3 hours
+    private Duration lockTimeoutDuration;
+
     public Optional<TransactionEntity> findById(String txId) {
         return transactionEntityRepository.findById(txId);
     }
 
-    public Set<TransactionEntity> findTransactionsByStatus(String organisationId, int pullTransactionsBatchSize) {
+    public Set<TransactionEntity> findAndLockTransactionsReadyToBeDispatched(String organisationId, int pullTransactionsBatchSize) {
         Set<BlockchainPublishStatus> dispatchStatuses = BlockchainPublishStatus.toDispatchStatuses();
         Limit limit = Limit.of(pullTransactionsBatchSize);
 
-        return transactionEntityRepository.findTransactionsByStatus(organisationId, dispatchStatuses, limit);
+        Set<TransactionEntity> transactionsByStatus = transactionEntityRepository.findTransactionsByStatus(
+                organisationId,
+                dispatchStatuses,
+                limit);
+        if (transactionsByStatus.isEmpty()) {
+            return transactionsByStatus;
+        }
+        // This logic could be moved to the repository, but for now it is easier to test it here
+        Set<TransactionEntity> filteredTransactions = transactionsByStatus.stream().filter(
+                transactionEntity -> (
+                        transactionEntity.getLockedAt()
+                                .map(lockedAt -> lockedAt.isBefore(LocalDateTime.now().minus(lockTimeoutDuration)))
+                                .orElse(true) // return true if lockedAt is not present
+                        ))
+                .collect(toSet());
+        filteredTransactions.forEach(tx -> tx.setLockedAt(LocalDateTime.now()));
```

Comment:
it is highly recommended to pass clock instance when dealing with time and tests